File: 099_Recover_Binary_Search_Tree/099_1.c
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

void inOrderTraverse(struct TreeNode *root, struct TreeNode **first, struct TreeNode **second, struct TreeNode **pre) {
    if (!root) return;
    inOrderTraverse(root->left, first, second, pre);
    if (!(*first) && *pre && root->val <= (*pre)->val) *first = *pre;
    if (*first && *pre && root->val <= (*pre)->val)  *second = root;
    *pre = root;
    inOrderTraverse(root->right, first, second, pre);
}

void recoverTree(struct TreeNode* root){
    struct TreeNode *first = NULL;
    struct TreeNode *second = NULL;
    struct TreeNode *pre = NULL;
    inOrderTraverse(root, &first, &second, &pre);
    int tmp = first->val;
    first->val = second->val;
    second->val = tmp;
}

File: 099_Recover_Binary_Search_Tree/099_1.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def recoverTree(self, root: TreeNode) -> None:
        first, second, pre = None, None, TreeNode(float("-inf"))
        stack = []
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if not first and root.val < pre.val:
                first = pre
            if first and root.val < pre.val:
                second = root
            pre = root
            root = root.right
        tmp = first.val
        first.val = second.val
        second.val = tmp